airbyte-cdk/python/airbyte_cdk/sources/connector_state_manager.py

```diff
@@ -14,41 +27,131 @@ class ConnectorStateManager:
     interface. It also provides methods to extract and update state
     """
 
-    # In the immediate, we only persist legacy which will be used during abstract_source.read(). In the subsequent PRs we will
-    # initialize the ConnectorStateManager according to the new per-stream interface received from the platform
-    def __init__(self, state: Union[List[AirbyteStateMessage], MutableMapping[str, Any]] = None):
-        if not state:
-            self.legacy = {}
-        elif self.is_migrated_legacy_state(state):
-            # The legacy state format received from the platform is parsed and stored as a single AirbyteStateMessage when reading
-            # the file. This is used for input backwards compatibility.
-            self.legacy = state[0].data
-        elif isinstance(state, MutableMapping):
-            # In the event that legacy state comes in as its original JSON object format, no changes to the input need to be made
-            self.legacy = state
-        else:
-            raise ValueError("Input state should come in the form of list of Airbyte state messages or a mapping of states")
+    def __init__(self, stream_instance_map: Mapping[str, Stream], state: Union[List[AirbyteStateMessage], MutableMapping[str, Any]] = None):
+        shared_state, per_stream_states = self._extract_from_state_message(state, stream_instance_map)
 
-    def get_stream_state(self, namespace: str, stream_name: str) -> AirbyteStateBlob:
-        # todo implement in upcoming PRs
-        pass
+        # We explicitly throw an error if we receive a GLOBAL state message that contains a shared_state because API sources are
+        # designed to checkpoint state independently of one another. API sources should never be emitting a state message where
+        # shared_state is populated. Rather than define how to handle shared_state without a clear use case, we're opting to throw an
+        # error instead and if/when we find one, we will then implement processing of the shared_state value.
+        if shared_state:
+            raise ValueError("API source connectors do not currently support GLOBAL AirbyteStateMessages that contain a shared_state")
```

Comment:
just to make it even clearer for the next 
```suggestion
            raise ValueError("Received a GLOBAL AirbyteStateMessages that contain a shared_state. But this library only ever generated per-STREAM STATE messages was not generated this connector. This must be an orchestrator or platform error. GLOBAL state messages with shared_state will not be processed correctly.")
```